File: android/app/src/test/java/com/dsview/player/ConfigTest.kt
package com.dsview.player

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class ConfigTest {

    @Test
    fun `clampInterval clampa abaixo do minimo`() {
        assertEquals(Config.SYNC_MIN, Config.clampInterval(0))
    }

    @Test
    fun `clampInterval clampa acima do maximo`() {
        assertEquals(Config.SYNC_MAX, Config.clampInterval(999999))
    }

    @Test
    fun `clampInterval valor dentro da faixa passa direto`() {
        assertEquals(120, Config.clampInterval(120))
    }

    @Test
    fun `buildRealApi null sem origin`() {
        assertNull(Config.buildRealApi("", "abc"))
    }

    @Test
    fun `buildRealApi null sem token`() {
        assertNull(Config.buildRealApi("https://site.com", ""))
    }

    @Test
    fun `buildRealApi monta a rota certa e tira a barra final do origin`() {
        assertEquals(
            "https://site.com.br/wp-json/ds-facil/v1/player/xyz",
            Config.buildRealApi("https://site.com.br/", "xyz"),
        )
    }

    @Test
    fun `buildRealApi URL-encoda o token`() {
        // java.net.URLEncoder é application/x-www-form-urlencoded: espaço vira "+", não "%20".
        assertEquals(
            "https://site.com/wp-json/ds-facil/v1/player/a+b%2Fc",
            Config.buildRealApi("https://site.com", "a b/c"),
        )
    }

    @Test
    fun `normalizeDomain vazio devolve vazio`() {
        assertEquals("", Config.normalizeDomain(""))
        assertEquals("", Config.normalizeDomain("   "))
    }

    @Test
    fun `normalizeDomain aceita dominio nu e assume https`() {
        assertEquals("https://suaempresa.com.br", Config.normalizeDomain("suaempresa.com.br"))
    }

    @Test
    fun `normalizeDomain preserva scheme e porta ja informados`() {
        assertEquals("http://site.com:8080", Config.normalizeDomain("http://site.com:8080"))
    }

    @Test
    fun `normalizeDomain tira caminho e barra final`() {
        assertEquals("https://site.com", Config.normalizeDomain("https://site.com/play/xxxx"))
        assertEquals("https://site.com", Config.normalizeDomain("site.com/"))
    }

    @Test
    fun `normalizeDomain string invalida devolve vazio`() {
        assertEquals("", Config.normalizeDomain("não é um domínio"))
    }

    // Bug real: "Sair" (ou desligar o auto-start) deixava o app preso num loop de relançamento —
    // por ser HOME do aparelho, o Android religava na hora, e a supressão antiga (igualdade simples
    // de boot-id, sem prazo) durava até reiniciar o TV box inteiro. Estes testes travam o
    // comportamento correto: suprime só a rajada instantânea, nunca uma reabertura de verdade.
    @Test
    fun `shouldSuppress logo apos o fechamento no mesmo boot suprime`() {
        val boot = 555L
        val quitAt = 10_000L
        assertEquals(true, Config.shouldSuppress(boot, quitAt, boot, quitAt + 1_000L))
    }

    @Test
    fun `shouldSuppress passada a janela nao suprime mesmo no mesmo boot`() {
        val boot = 556L
        val quitAt = 10_000L
        val muitoDepois = quitAt + Config.AUTO_RELAUNCH_SUPPRESS_MS + 1_000L
        assertEquals(false, Config.shouldSuppress(boot, quitAt, boot, muitoDepois))
    }

    @Test
    fun `shouldSuppress boot diferente nunca suprime`() {
        assertEquals(false, Config.shouldSuppress(777L, 10_000L, 778L, 10_500L))
    }

    @Test
    fun `shouldSuppress nunca fechou de proposito nao suprime`() {
        assertEquals(false, Config.shouldSuppress(0L, 0L, 999L, 999_999L))
    }

    @Test
    fun `shouldSuppress config legado sem quitAt nao suprime`() {
        assertEquals(false, Config.shouldSuppress(42L, 0L, 42L, 100L))
    }
}

File: windows/test/config.test.js
'use strict';
require('./_stub-electron');
const test = require('node:test');
const assert = require('node:assert/strict');
const fs = require('fs');
const path = require('path');
const config = require('../src/main/config');

test('syncIntervalMin: clampa abaixo do mínimo (5)', () => {
  assert.equal(config.syncIntervalMin({ syncInterval: 1 }), 5);
});

test('syncIntervalMin: clampa acima do máximo (1440)', () => {
  assert.equal(config.syncIntervalMin({ syncInterval: 999999 }), 1440);
});

test('syncIntervalMin: valor dentro da faixa passa direto', () => {
  assert.equal(config.syncIntervalMin({ syncInterval: 120 }), 120);
});

test('syncIntervalMin: não-numérico cai no padrão (60), não no mínimo', () => {
  assert.equal(config.syncIntervalMin({ syncInterval: 'abacate' }), config.SYNC_DEFAULT);
  assert.equal(config.syncIntervalMin({ syncInterval: undefined }), config.SYNC_DEFAULT);
});

test('realApi: null sem origin ou sem token', () => {
  assert.equal(config.realApi({ origin: '', token: 'abc' }), null);
  assert.equal(config.realApi({ origin: 'https://site.com', token: '' }), null);
});

test('realApi: monta a rota REST certa e tira a barra final do origin', () => {
  assert.equal(
    config.realApi({ origin: 'https://site.com.br/', token: 'xyz' }),
    'https://site.com.br/wp-json/ds-facil/v1/player/xyz',
  );
});

test('realApi: URL-encoda o token', () => {
  assert.equal(
    config.realApi({ origin: 'https://site.com', token: 'a b/c' }),
    'https://site.com/wp-json/ds-facil/v1/player/a%20b%2Fc',
  );
});

test('write/read: grava e lê de volta (round-trip via arquivo real)', () => {
  const written = config.write({ origin: 'https://x.com', token: 't1', favorites: [{ name: 'A', url: 'u' }] });
  assert.equal(written.origin, 'https://x.com');
  const read = config.read();
  assert.equal(read.origin, 'https://x.com');
  assert.equal(read.token, 't1');
  assert.deepEqual(read.favorites, [{ name: 'A', url: 'u' }]);
});

test('write: escreve atômico — não sobra .tmp depois de gravar', () => {
  config.write({ origin: 'https://atomic-test.com' });
  assert.equal(fs.existsSync(config.FILE + '.tmp'), false);
  assert.equal(fs.existsSync(config.FILE), true);
});

test('write: patch parcial preserva os outros campos', () => {
  config.write({ origin: 'https://preserva.com', token: 'tk', autostart: true });
  config.write({ autostart: false });
  const read = config.read();
  assert.equal(read.origin, 'https://preserva.com');
  assert.equal(read.token, 'tk');
  assert.equal(read.autostart, false);
});

test('read: JSON corrompido no disco cai nos DEFAULTS em vez de lançar', () => {
  fs.mkdirSync(path.dirname(config.FILE), { recursive: true });
  fs.writeFileSync(config.FILE, '{ isso não é json válido');
  const read = config.read();
  assert.equal(read.origin, '');
  assert.equal(read.offline, true);
});

test('normalizeDomain: vazio devolve vazio', () => {
  assert.equal(config.normalizeDomain(''), '');
  assert.equal(config.normalizeDomain('   '), '');
});

test('normalizeDomain: aceita domínio nu e assume https', () => {
  assert.equal(config.normalizeDomain('suaempresa.com.br'), 'https://suaempresa.com.br');
});

test('normalizeDomain: preserva scheme e porta já informados', () => {
  assert.equal(config.normalizeDomain('http://site.com:8080'), 'http://site.com:8080');
});

test('normalizeDomain: tira caminho e barra final', () => {
  assert.equal(config.normalizeDomain('https://site.com/play/xxxx'), 'https://site.com');
  assert.equal(config.normalizeDomain('site.com/'), 'https://site.com');
});

test('normalizeDomain: string inválida devolve vazio', () => {
  assert.equal(config.normalizeDomain('não é um domínio'), '');
});
